portal/_extensions/truncatehtml.py

```diff
@@ -0,0 +1,179 @@
+            self.counter += 1
+            if char == '&':
+                return self.__entity()
+            elif char != '<':
+                return char
+            elif self.input[self.counter] == '/':
+                self.counter += 1
+                return self.__close_tag()
+            else:
+                return self.__open_tag()
+        except IndexError:
+            return END
+
+    def __entity(self):
+        """Return a token representing an HTML character entity.
+        Precondition: self.counter points at the charcter after the &
+        Postcondition: self.counter points at the character after the ;
+        """
+        char = self.input[self.counter]
+        entity = ['&']
+        while char != ';':
+            entity.append(char)
+            char = self.__next_char()
+        entity.append(';')
+        self.counter += 1
+        return ''.join(entity)
+
+    def __open_tag(self):
+        """Return an open/close tag token.
+        Precondition: self.counter points at the first character of the tag name
+        Postcondition: self.counter points at the character after the <tag>
+        """
+        char = self.input[self.counter]
+        tag = []
+        rest = []
+        while char != '>' and char != ' ':
+            tag.append(char)
+            char = self.__next_char()
+        while char != '>':
+            rest.append(char)
+            char = self.__next_char()
+        if self.input[self.counter - 1] == '/':
+            self.counter += 1
+            return SelfClosingTag(''.join(tag), ''.join(rest))
+        elif ''.join(tag) in VOID_ELEMENTS:
+            self.counter += 1
+            return SelfClosingTag(''.join(tag), ''.join(rest))
+        else:
+            self.counter += 1
+            return OpenTag(''.join(tag), ''.join(rest))
+
+    def __close_tag(self):
+        """Return an open/close tag token.
+        Precondition: self.counter points at the first character of the tag name
+        Postcondition: self.counter points at the character after the <tag>
+        """
+        char = self.input[self.counter]
+        tag = []
+        while char != '>':
+            tag.append(char)
+            char = self.__next_char()
+        self.counter += 1
+        return CloseTag(''.join(tag))
+
+
+def truncate(str, target_len, ellipsis=''):
+    """Returns a copy of str truncated to target_len characters,
+    preserving HTML markup (which does not count towards the length).
+    Any tags that would be left open by truncation will be closed at
+    the end of the returned string.  Optionally append ellipsis if
+    the string was truncated."""
+    stack = []  # open tags are pushed on here, then popped when the matching close tag is found
+    retval = []  # string to be returned
+    length = 0  # number of characters (not counting markup) placed in retval so far
+    tokens = Tokenizer(str)
+    tok = tokens.next_token()
+    while tok != END:
+        if not length < target_len:
```

Comment:
Changing the termination condition to the following will truncate only on blank spaces "near" the target character length:

```suggestion
        if length >= target_len and tok == ' ':
```